proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,354 @@
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma vgic_update_cur_vcpu_valid[wp]:
+  "vgic_update vr f \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: vgic_update_def vcpu_update_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma vcpu_disable_cur_vcpu_valid[wp]:
+  "vcpu_disable vo \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: vcpu_disable_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply wpsimp
+  done
+
+crunches do_machine_op
+  for arch_tcb_at_cur_thread[wp]: "\<lambda>s. arch_tcb_at P (cur_thread s) s"
+  (wp: cur_vcpu_valid_lift crunch_wps simp: active_cur_vcpu_of_def simp: crunch_simps)
+
+lemma set_vcpu_arch_tcb_at_cur_thread[wp]:
+  "set_vcpu ptr vcpu \<lbrace>\<lambda>s. arch_tcb_at P (cur_thread s) s\<rbrace>"
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+crunches vcpu_disable, vcpu_restore, vcpu_save_reg_range, vgic_update_lr, vcpu_save, set_vm_root
+  for arch_tcb_at_cur_thread[wp]: "\<lambda>s. arch_tcb_at P (cur_thread s) s"
+  (wp: crunch_wps simp: active_cur_vcpu_of_def simp: crunch_simps active_cur_vcpu_of_def)
+
+crunches vcpu_update, do_machine_op, invalidate_hw_asid_entry, invalidate_asid
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  (wp: crunch_wps simp: active_cur_vcpu_of_def simp: crunch_simps active_cur_vcpu_of_def)
+
+crunches vcpu_restore, do_machine_op, vcpu_save_reg, vgic_update, save_virt_timer
+  for cur_vcpu_valid[wp]: cur_vcpu_valid
+  (wp: cur_vcpu_valid_lift crunch_wps mapM_x_wp_inv simp: active_cur_vcpu_of_def crunch_simps)
+
+lemma vcpu_save_reg_active_cur_vcpu_of[wp]:
+  "vcpu_save_reg a b \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  by (wpsimp simp: vcpu_save_reg_def)
+
+crunches vcpu_save_reg_range, vcpu_save_reg_range, vgic_update_lr, vcpu_enable, vcpu_save
+  for cur_vcpu_valid[wp]: cur_vcpu_valid
+  (wp: cur_vcpu_valid_lift crunch_wps mapM_x_wp_inv simp: active_cur_vcpu_of_def crunch_simps)
+
+lemma switch_vcpu_cur_vcpu_valid[wp]:
+  "\<lbrace>\<lambda>s. arch_tcb_at (\<lambda>a. itcb_vcpu a = v) t s\<rbrace>
+   vcpu_switch v
+   \<lbrace>\<lambda>_ s. cur_vcpu_valid (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  apply (clarsimp simp: vcpu_switch_def)
+  apply (cases v; clarsimp)
+   apply (rule hoare_seq_ext[OF _ gets_sp], rename_tac cur_vcpu)
+   apply (case_tac cur_vcpu; clarsimp)
+    apply (wpsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def)
+   apply (clarsimp simp: when_def)
+   apply (intro conjI impI)
+    apply (rule hoare_seq_ext_skip, wpsimp)
+    apply (wpsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def)
+   apply (wpsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def)
```

Comment:
same comment as below, something about this feels... very repetitive; it might be necessary, but looking at it again couldn't hurt. Hard to say without seeing the proof states, happy to be wrong here.